spongeauth/accounts/views.py

```diff
@@ -43,6 +43,20 @@ class ForgotTokenGenerator(django.contrib.auth.tokens.PasswordResetTokenGenerato
 
 
 def _log_user_in(request, user, skip_twofa=False):
+    # Resync groups with the TOS acceptances.
+    # XXX(lukegb): this is a hack, don't do this.
```

Comment:
Is this syncing the groups in memory with the groups in the DB? Do you want to put this into a separate function and have these sync periodically?

Reply:
Specifically this syncs the group membership in the *database* with the ToS acceptances. These are separate because going back and retroactively changing the group membership to use a through table is tricky, and acceptances store a timestamp value as well, whereas groups do not.

This should really be a trigger when acceptances change rather than on login, but I haven't bothered to implement that yet.

This probably has consequences for the following flow, though:

1. User logs in, and hasn't accepted all the ToSes yet
2. User clicks through the ToS acknowledgement
3. User continues to Discourse/Ore

...because the user *won't* be added to the ToS group for the things they've just acked.